src/librustc_span/symbol.rs

```diff
@@ -929,6 +924,53 @@ impl UseSpecializedDecodable for Ident {
     }
 }
 
+/// This is the most general way to print identifiers.
+/// AST pretty-printer is used as a fallback for turning AST structures into token streams for
+/// proc macros. Additionally, proc macros may stringify their input and expect it survive the
+/// stringification (especially true for proc macro derives written between Rust 1.15 and 1.30).
+/// So we need to somehow pretty-print `$crate` in a way preserving at least some of its
+/// hygiene data, most importantly name of the crate it refers to.
+/// As a result we print `$crate` as `crate` if it refers to the local crate
+/// and as `::other_crate_name` if it refers to some other crate.
+/// Note, that this is only done if the ident token is printed from inside of AST pretty-pringing,
+/// but not otherwise. Pretty-printing is the only way for proc macros to discover token contents,
+/// so we should not perform this lossy conversion if the top level call to the pretty-printer was
+/// done for a token stream or a single token.
+pub struct IdentPrinter {
+    symbol: Symbol,
+    is_raw: bool,
+    convert_dollar_crate: Option<Span>,
+}
+
+impl IdentPrinter {
+    pub fn new(symbol: Symbol, is_raw: bool, convert_dollar_crate: Option<Span>) -> IdentPrinter {
```

Comment:
This basically appears to be for `token_kind_to_string_ext` only -- where does that get it's span from? I suspect that in practice almost everyone should *not* be calling this, but prefer from_ident instead. Could we either rename this to a less innocuous sounding name than new (perhaps "with_crate_span"?) or add a comment saying "prefer from_ident" here?

Reply:
The span is the identifier's span, it either comes from `token.span` for identifier tokens or from `ident.span` for AST idents.

Neither `new` nor `from_ident` should be called by anyone, most likely.
Printing functions for `token::(Nt)Ident` and `ast::Ident` already call them and that should be enough.

Reply:
Makes sense. Could we document that (prefer the printing functions for ast/tokens)?